pysmt/optimization/optimizer.py

```diff
@@ -138,6 +140,93 @@ def _get_lt(self, objective_formula):
             raise PysmtValueError("Invalid optimization function type: %s" % otype)
 
 
+class OptSearchInterval():
+
+    def __init__(self, goal):
+        self._obj = goal
+        self._lower = None #-INF where i found this costant?
+        self._upper = None #+INF
+        self._pivot = None
+
+    def _comparation_functions(self, objective_formula):
+        """Internal utility function to get the proper cast, LT and LE
+        function for the given objective formula
+        """
+        otype = self.environment.stc.get_type(objective_formula)
+        mgr = self.environment.formula_manager
+        if otype.is_int_type():
+            return mgr.Int, mgr.LT, mgr.LE
+        elif otype.is_real_type():
+            return mgr.Real, mgr.LT, mgr.LE
+        elif otype.is_bv_type():
+            return (lambda x: mgr.BV(x, otype.width)), mgr.BVULT, mgr.BVULE
+        else:
+            raise PysmtValueError("Invalid optimization function type: %s" % otype)
+
+    def _make_better_than(self, goal, x):
+        cast, lt, _ = self._comparation_functions(goal.term())
+        if goal.is_minimization_goal():
+            t1 = goal.term()
+            t2 = x
+        else:
+            t1 = x
+            t2 = goal.term()
+        return lt(t1, t2)
+
+    def linear_search_cut(self):
+        """must be called always"""
+        if self._obj.is_minimization_goal():
+            bound = self._upper
```

Comment:
Or it can just remember the last model value (see changes in the 2nd email).
Either is fine to me.